### PR TITLE
Enhance `_get_and_store_range` Function to Include Trip Statistics and Last API Call Tracking

### PR DESCRIPTION
**Pull request by TeachMeTW**
### Overview
Enhances `_get_and_store_range` by adding `total_trips`, `labeled_trips`, and `last_call` fields to the user profile, enabling better insights into user activities and API usage.

### Changes

1. **Extended User Profile Updates:**
   - **`total_trips`**: Total number of trips recorded for the user.
   - **`labeled_trips`**: Number of labeled/annotated trips.

2. **Integrated Last API Call Tracking:**
   - Accesses `stats/server_api_time` to retrieve timestamps of the last `GET` and `PUT` API calls.
   - Determines the latest call and stores it as `last_call` in the user's profile in UNIX format.

### Testing

1. Picked a user, cleared pipeline, re-ran pipeline, verified user profile was updated.

**Please review and provide feedback or approval.**

### CODE REVIEW

**Review by JGreenlee — CHANGES_REQUESTED** [review]

File: emission/pipeline/intake_stage.py
Comment (TeachMeTW) [reply emission/pipeline/intake_stage.py:248]: @JGreenlee I get this error:
`extra query would overwrite keys ['user_id']` using op-admins implementation

Comment (JGreenlee) [reply emission/pipeline/intake_stage.py:248]: You can remove the `user_id` filter because `ts` here is already the timeseries **for that user**

**Comment by TeachMeTW** [comment]
@JGreenlee Comments addressed, I split the function into a different file as suggested -- though it may be overkill. Seeking thoughts

**Review by JGreenlee — COMMENTED** [review]
Good refactor! I agree that the smaller functions are overkill.
Sometimes it's just a matter of finding the balance between tidiness and complexity.

**Comment by TeachMeTW** [comment]
@JGreenlee Comments addressed

File: emission/analysis/result/user_stat.py
Comment (JGreenlee) [inline emission/analysis/result/user_stat.py:104]: Are we forgetting to store `last_call_ts` ?

**Comment by TeachMeTW** [comment]
@JGreenlee Addressed and rebased

File: emission/tests/analysisTests/intakeTests/TestUserStat.py
Comment (JGreenlee) [inline emission/tests/analysisTests/intakeTests/TestUserStat.py:127]: If we want to validate last_call_ts getting filled in, we will need to pretend we accessed the API. I think you can do this by calling `after_request()` from `emission.net.api.cfc_webapp`

I recommend extracting this to a different function `test_last_call_ts`.

After calling `after_request()`, `runIntakePipeline` again. Then get the profile and `last_call_ts` should be filled in with the time at which you called `after_request()`

**Comment by TeachMeTW** [comment]
@JGreenlee I realized my code was NOT pushed and you may have reviewed the wrong thing.... my bad, I will take your notes into account with the latest.

File: emission/tests/analysisTests/intakeTests/TestUserStat.py
Comment (TeachMeTW) [reply emission/tests/analysisTests/intakeTests/TestUserStat.py:104]: @JGreenlee load any user inputs how so? Like loading from the example or what?

**Comment by TeachMeTW** [comment]
@JGreenlee Comments addressed, ready for re-review

File: emission/tests/analysisTests/intakeTests/TestUserStat.py
Comment (JGreenlee) [inline emission/tests/analysisTests/intakeTests/TestUserStat.py:35]: can you move this to the end of file below `if __name__ == '__main__':` ?

File: emission/tests/analysisTests/intakeTests/TestUserStat.py
Comment (JGreenlee) [inline emission/tests/analysisTests/intakeTests/TestUserStat.py:49]: Just calling `setupRealExample` (not `setupRealExampleWithEntries`) will do all of this for you

It will create one UUID for you, assigned to `self.testUUID`. Use that instead of `self.UUID`

File: emission/tests/analysisTests/intakeTests/TestUserStat.py
Comment (JGreenlee) [inline emission/tests/analysisTests/intakeTests/TestUserStat.py:55]: What happens if you take this out? Does something not work without it?

File: emission/tests/analysisTests/intakeTests/TestUserStat.py
Comment (JGreenlee) [inline emission/tests/analysisTests/intakeTests/TestUserStat.py:80]: This can be more concise, no need to overexplain. I would simply do:

```python
edb.get_timeseries_db().delete_many({"user_id": self.UUID})
edb.get_pipeline_state_db().delete_many({"user_id": self.UUID})
edb.get_analysis_timeseries_db().delete_many({"user_id": self.UUID})
edb.get_profile_db().delete_one({"user_id": self.UUID})
```

File: emission/tests/analysisTests/intakeTests/TestUserStat.py
Comment (JGreenlee) [inline emission/tests/analysisTests/intakeTests/TestUserStat.py:82]: this test looks good

File: emission/tests/analysisTests/intakeTests/TestUserStat.py
Comment (JGreenlee) [inline emission/tests/analysisTests/intakeTests/TestUserStat.py:123]: This ts could be anything, but using the same ts as the end_ts could be confusing to people in the future.
You could grab the current time and call it `test_call_ts`

```suggestion
        test_call_ts = time.time()
        enac.store_server_api_time(self.UUID, "test_call", test_call_ts)
```


Then you can also use `test_call_ts` below instead of declaring `expected_last_call_ts`

**Review by JGreenlee — CHANGES_REQUESTED** [review]
Almost there!

File: emission/tests/analysisTests/intakeTests/TestUserStat.py
Comment (JGreenlee) [reply emission/tests/analysisTests/intakeTests/TestUserStat.py]: The `real_examples` files like `shankari_2015-aug-27` only have raw sensed data; i.e. what the phone automatically picked while traveling

For some of the days, e.g. `shankari_2016-06-20` we have an additional file `shankari_2016-06-20.user_inputs`, which does not contain sensed data but it has "manual inputs" or "user inputs". These entries represent labels or survey responses that the user has recorded while using the app.

For a more in-depth test you could load both `shankari_2016-06-20` and `shankari_2016-06-20.user_inputs` and `expected_labeled_trips` would update

File: emission/tests/analysisTests/intakeTests/TestUserStat.py
Comment (TeachMeTW) [reply emission/tests/analysisTests/intakeTests/TestUserStat.py:35]: The setup or  just `enc.configLogging()` ?

File: emission/tests/analysisTests/intakeTests/TestUserStat.py
Comment (TeachMeTW) [reply emission/tests/analysisTests/intakeTests/TestUserStat.py:49]: Noted, that does seem way simpler

File: emission/tests/analysisTests/intakeTests/TestUserStat.py
Comment (TeachMeTW) [reply emission/tests/analysisTests/intakeTests/TestUserStat.py:55]: When I take this out, the profile is empty.
```
Traceback (most recent call last):
  File "/Users/rsimpson/e-mission-server/emission/tests/analysisTests/intakeTests/TestUserStat.py", line 81, in testGetAndStoreUserStats
    self.assertIsNotNone(profile, "User profile should exist after storing stats.")
AssertionError: unexpectedly None : User profile should exist after storing stats.
```

Meaning for some reason the entry is not being inserted. Does that hint to an issue with my additions or existing?

**Comment by TeachMeTW** [comment]
@JGreenlee Addressed, see response -- in need of clarification

File: emission/tests/analysisTests/intakeTests/TestUserStat.py
Comment (JGreenlee) [reply emission/tests/analysisTests/intakeTests/TestUserStat.py:35]: I just meant configLogging 
The other tests have it at the bottom (at least the majority)

File: emission/analysis/result/user_stat.py
Comment (JGreenlee) [inline emission/analysis/result/user_stat.py:12]: nitpick: I had an earlier comment here that was resolved but I don't think was addressed. I don't see the point of this function when it is just a wrapper around `ts.find_entries_count` with the exact same params and no added functionality 

In the event that a suggested change doesn't work for you or you have a good reason to do otherwise, that's fine; just reply and we can discuss

Or, if you just missed my comment, make sure not to resolve conversations until you have made a change that addresses them. When I have my code reviewed I like to react with thumbs up and reply "TODO" so I can keep track of what I have seen vs what I have actually implemented and pushed up

**Review by JGreenlee — CHANGES_REQUESTED** [review]
@TeachMeTW Code looks good but the new tests appear to be failing in GH Actions 
Do they pass for you locally?

File: emission/analysis/result/user_stat.py
Comment (TeachMeTW) [reply emission/analysis/result/user_stat.py]: I think I did address it, but I've been trying to rebase and do all this github branch magic and maybe it got reverted -- will do react and/or reply on future comments before resolving

OR I believe a better option is for the reviewer to resolve the comment? what are your thoughts on that approach in addition?

**Comment by JGreenlee** [comment]
I don't know why it's passing locally and not here, but it is related to this PR because the test that's failing is the one you added. It's complaining about timestamps that look a little bit off.

**Comment by TeachMeTW** [comment]
> I don't know why it's passing locally and not here, but it is related to this PR because the test that's failing is the one you added. It's complaining about timestamps that look a little bit off.

Shall I add delta/assert if almost equal?

**Comment by JGreenlee** [comment]
> > I don't know why it's passing locally and not here, but it is related to this PR because the test that's failing is the one you added. It's complaining about timestamps that look a little bit off.
> 
> Shall I add delta/assert if almost equal?

No we should investigate why it's different.
The pipeline is supposed to be deterministic, i.e. 100% reproducible 

What happens if you checkout master and run that test data through the pipeline?
Is it the same as the CI or the same as you got on your branch?

**Comment by TeachMeTW** [comment]
@JGreenlee  I checkout master and run the test data through pipeline and its the same as the branch NOT the CI. I believe running ALL the tests must affect the ts. I tried to test runAllTest.sh but it crashes my console/vscode, perhaps you could verify

**Comment by JGreenlee** [comment]
I will try to verify tomorrow. I will also pass this along in the review process for Shankari's input.

If `runAllTests` performs differently than running an individual test file, it is concerning as it means we do not have proper isolation between the tests. Maybe one of the tests does not `tearDown` properly. It may be difficult to track down the culprit.

**Comment by JGreenlee** [comment]
If we can confirm that the issue is with one of the other tests, this looks good to me to merge.

**Comment by shankari** [comment]
> I tried to test runAllTest.sh but it crashes my console/vscode, perhaps you could verify

You should resolve the crash and then test. We already have one CI failing for "unknown reasons", having the second also fail makes the CI not be meaningful.

**Comment by TeachMeTW** [comment]
@shankari  @JGreenlee  

I discovered two things.

### Crashing
The first one in regards to the crashing runAllTests.sh

When the terminal started crashing during script execution, I redirected the output to a log file to figure out what was going on. That’s when I saw the error: ValueError: I/O operation on closed file. Initially, the code opened the file without using a with statement and didn’t explicitly close it causing a different ValueError where the file was never closed. This was causing unpredictable behavior, as the file could be closed unexpectedly by the system or Python's garbage collector.

I updated the code to use a with statement to ensure the file was properly closed. However, the error persisted because there was still a redundant line after the with block trying to read the file, even though it had already been closed. Once I removed that redundant line and ensured the file was read entirely within the with block, the issue was resolved, and the terminal stopped crashing.

I could make a new issue in regards to this fix.

### Timing differences

There is indeed a difference with single testing and runAll. The single test has the ts normal but the runAll brings the same error as the gitlab runner where it is off. I thought fixing the crash would fix this discrepancy but it seems deeper than that.

**Comment by shankari** [comment]
> My hunch is that a teardown is broken somewhere and this UUID is being reused. Perhaps I can use a different uuid as a bandaid fix or can dig deeper into the root cause.

Feel free to use either option. I will not merge until this is resolved.

**Comment by JGreenlee** [comment]
This specific problem has happened before.

I used Github search to search the `e-mission` org for the problematic timestamp and got exactly 1 hit https://github.com/e-mission/e-mission-server/pull/509#issuecomment-367734717

Reading through that PR, we cannot pinpoint what the cause was.

**Comment by TeachMeTW** [comment]
Commenting out all the test cases in TestTripSegmentation and replacing it with a dummy test with a print statment still invokes the same errors. Might be something with setup and teardown.

**Comment by TeachMeTW** [comment]
However, if one runs `TestTripSegmentation` or `TestSectionSegmentation` **AFTER** `TestGetUserStats` there is **NO ERROR**

**Comment by TeachMeTW** [comment]
Alternatively, using `emission/tests/data/real_examples/shankari_2015-aug-21` rather than `emission/tests/data/real_examples/shankari_2015-aug-27` resulted in **NO ERRORS** when running all.

image

<br></br>

We have a couple options. 

1. We could do above and use Aug 21 dataset for UserStats, call it a day for now.
2. We could dig deeper as to why this is happening.

Thoughts?

**Comment by shankari** [comment]
We should do both.
We should change the test to Aug 21 for UserStats so that we can merge this and continue to make progress.
We should then file a new issue to dig deeper into why this is happening

**Comment by TeachMeTW** [comment]
@shankari Modified and pushed accordingly. Will make an issue as suggested.

**Comment by JGreenlee** [comment]
> Will make an issue as suggested.

@TeachMeTW I figured out what was causing it and made a PR https://github.com/e-mission/e-mission-server/pull/1003

**Comment by TeachMeTW** [comment]
Nice find!

**Review by shankari — APPROVED** [review]
I am going to roll forward and merge to unblock us; @TeachMeTW please address these comments in a follow-on PR

**Comment by shankari** [comment]
@TeachMeTW there is now a merge conflict after I have pulled from master.
Screenshot 2024-12-21 at 4 58 28 PM

I think this may be because your original base did not include 
https://github.com/e-mission/e-mission-server/commit/2a327421d20b4dd5e3989cdd042df9e645bd1707

You need to pull changes from master periodically, particularly when you are making changes to the same files

**Comment by shankari** [comment]
@JGreenlee @TeachMeTW this needs to be refactored further for it to work properly.
Note that we have an early return if there is no new data, so this code is effectively in the "else"

```
        if skip_if_no_new_data and new_entry_count == 0:
            print("No new entries, and skip_if_no_new_data = %s, skipping the rest of the pipeline" % skip_if_no_new_data)
            return
        else:
            print("New entry count == %s and skip_if_no_new_data = %s, continuing" %
                (new_entry_count, skip_if_no_new_data))
```

That's fine as long as we were only caching pipeline state, because it wouldn't change if the pipeline was skipped
But now that we are caching the `last_call_ts` the current location will not work.
Concretely, if the app is contacting the server but is not uploading any data, `last_call_ts` will never be changed.

**Comment by shankari** [comment]
I have fixed this in #1005, @TeachMeTW please update the unit tests as outlined there

File: emission/analysis/result/user_stat.py
Comment (TeachMeTW) [reply emission/analysis/result/user_stat.py:66]: will be addressing in follow up pr; no reason it should be hard coded

**Pull request merged**
